test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/AtomicCryptoTransferHTSSuite.java

```diff
@@ -0,0 +1,1308 @@
+import com.hedera.services.legacy.proto.utils.ByteStringUtils;
+import com.hederahashgraph.api.proto.java.TokenSupplyType;
+import com.hederahashgraph.api.proto.java.TokenType;
+import java.util.List;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class AtomicCryptoTransferHTSSuite extends HapiApiSuite {
+    private static final Logger log = LogManager.getLogger(AtomicCryptoTransferHTSSuite.class);
+
+    private static final Tuple[] EMPTY_TUPLE_ARRAY = new Tuple[] {};
+    private static final long GAS_TO_OFFER = 5_000_000L;
+    private static final long TOTAL_SUPPLY = 1_000;
+    private static final String NFT_TOKEN = "AToken_NFT";
+    private static final String TOKEN_TREASURY = "the_treasury";
+    private static final String FUNGIBLE_TOKEN = "AToken";
+    private static final String RECEIVER = "the_receiver";
+    private static final String RECEIVER2 = "the_receiver2";
+    private static final String SENDER = "the_sender";
+    private static final String SENDER2 = "the_sender2";
+    private static final String DELEGATE_KEY = "contractKey";
+    private static final String MULTI_KEY = "purpose";
+    private static final String OWNER = "Owner";
+    private static final KeyShape DELEGATE_CONTRACT_KEY_SHAPE =
+            KeyShape.threshOf(1, KeyShape.SIMPLE, DELEGATE_CONTRACT);
+    private static final String CONTRACT = "AtomicCryptoTransfer";
+    private static final String TRANSFER_MULTIPLE_TOKENS = "transferMultipleTokens";
+    private static final String BASE_APPROVAL_TXN = "baseApproveTxn";
+
+    public static final String SECP_256K1_SOURCE_KEY = "secp256k1Alias";
+
+    public static void main(String... args) {
+        new AtomicCryptoTransferHTSSuite().runSuiteAsync();
+    }
+
+    @Override
+    public boolean canRunConcurrent() {
+        return true;
+    }
+
+    @Override
+    public List<HapiApiSpec> getSpecsInSuite() {
+        return List.of(
+                new HapiApiSpec[] {
+                    cryptoTransferForHbarOnly(),
+                    cryptoTransferForFungibleTokenOnly(),
+                    cryptoTransferForNonFungibleTokenOnly(),
+                    cryptoTransferHBarFungibleNft(),
+                    cryptoTransferAllowanceHbarToken(),
+                    cryptoTransferAllowanceFungibleToken(),
+                    cryptoTransferAllowanceNft(),
+                    cryptoTransferSpecialAccounts()
+                });
+    }
+
+    private HapiApiSpec cryptoTransferForHbarOnly() {
+        final var cryptoTransferTxn = "cryptoTransferTxn";
+        final var cryptoTransferMultiTxn = "cryptoTransferMultiTxn";
+        final var cryptoTransferRevertTxn = "cryptoTransferRevertTxn";
+        final var cryptoTransferRevertNoKeyTxn = "cryptoTransferRevertNoKeyTxn";
+        final var cryptoTransferRevertBalanceTooLowTxn = "cryptoTransferRevertBalanceTooLowTxn";
+
+        return defaultHapiSpec("cryptoTransferForHbarOnly")
```

Comment:
I see that this spec (and probably the others, haven't gotten that far yet) have brief inline comments describing the test being performed _but_ those lines are widely spread apart and, actually, so "embedded" into a wall of code that they're hard to spot.

I would recommend each spec also contain a comment block describing in _one place_ (at the top) what exactly the subtests being performed are.  An indexed list.  Those descriptions can be very brief.  And _then_ each place there is _currently_ a brief comment put a more extensive comment like "2 parties with 1000hbar each, transferring 200hbar, succeeds", or if it fails, including _why_ it should have failed "insufficient balance" etc. (more informative than `CONTRACT_REVERT_EXECUTED`).

The reason I suggest this is that these tests are going to be _read_ at the point when something has gone _wrong_ - a stressful situation in the first place - and the _reader_ is not going to be the _author_ and will want to have a good concise but detailed description of _specifically_ what each subtest is trying to do - rather than trying to decipher it from a wall of code _only_.  (And even if the troubleshooter/reader _is_ the author ... if it is more than 6months from now ... will you _still_ remember what _each_ test case is for?). In my view each of this test subcases is like a "twisty maze of passages, _all similar_" and trying to decipher what's going on with a particular failure - when under time pressure - is likely to be frustrating (and possibly error prone).

Reply:
I'm in agreement with large parts of this comment.  I think there are probably things we can do to simplify integration/end-to-end creation.  Perhaps we can gather some ideas together and document as a separate issue and address in another pr.

In the meantime, I think your comment about using a comment block is reasonable.  I will update in cases when more than one test case is performed in a single test method.

Reply:
Comments are great!